pkg/network/setup/controller_test.go

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network_test
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "kubevirt.io/api/core/v1"
+	netsetup "kubevirt.io/kubevirt/pkg/network/setup"
+)
+
+var _ = Describe("controller", func() {
+	var netCtrl netsetup.Controller
+	var vmi *v1.VirtualMachineInstance
+
+	BeforeEach(func() {
+		netCtrl = netsetup.NewController(nil)
+
+		vmi = &v1.VirtualMachineInstance{}
+		vmi.UID = "123"
+	})
+
+	It("runs setup successfully", func() {
+		Expect(netCtrl.Setup(vmi, 0, doNetNSDummyNoop, netPreSetupDummyNoop)).To(Succeed())
+		Expect(netCtrl.SetupCompleted(vmi)).To(BeTrue())
+	})
+
+	It("runs teardown successfully", func() {
+		Expect(netCtrl.Setup(vmi, 0, doNetNSDummyNoop, netPreSetupDummyNoop)).To(Succeed())
+		Expect(netCtrl.SetupCompleted(vmi)).To(BeTrue())
+		Expect(netCtrl.Teardown(vmi, netTeardownDummyNoop)).To(Succeed())
+		Expect(netCtrl.SetupCompleted(vmi)).To(BeFalse())
+	})
+
+	It("skips secondary setup runs", func() {
+		Expect(netCtrl.Setup(vmi, 0, doNetNSDummyNoop, netPreSetupDummyNoop)).To(Succeed())
+		Expect(netCtrl.Setup(vmi, 0, doNetNSDummyNoop, netPreSetupFail)).To(Succeed())
+		Expect(netCtrl.SetupCompleted(vmi)).To(BeTrue())
+	})
+
+	It("fails the pre-setup run", func() {
+		Expect(netCtrl.Setup(vmi, 0, doNetNSDummyNoop, netPreSetupFail)).NotTo(Succeed())
+		Expect(netCtrl.SetupCompleted(vmi)).To(BeFalse())
+	})
+
+	It("fails the setup run", func() {
+		Expect(netCtrl.Setup(vmi, 0, doNetNSFail, netPreSetupDummyNoop)).NotTo(Succeed())
```

Comment:
I would prefer to have this `doNetNSFail` func locally defined within the test, but I won't insist. The test file is small enough for me not to have to fish around for the definition.